### PR TITLE
More rules logging and L003 indent improvement.

### LINKED ISSUE #734: sqlfluff fix uses incorrect indentation for `USING` keyword
-- If this is a parsing or linting issue, please include a minimal SQL example which reproduces the issue, along with the `sqlfluff parse` output, `sqlfluff lint` output and `sqlfluff fix` output when relevant. Finally, please add any labels that are related to the issue before submitting so that it can be addressed easily! --
`sqlfluff fix` seems to indent the `USING` keyword three levels deep.

```
SELECT
  *
FROM foo
INNER JOIN bar
  USING (baz)
WHERE qux = 1
```
This code has two areas with 2 space indentation, and those should be fixed accordingly.

## Expected Behaviour
```
SELECT
    *
FROM foo
INNER JOIN bar
    USING (baz)
WHERE qux = 1
```
## Observed Behaviour
```
SELECT
    *
FROM foo
INNER JOIN bar
           USING (baz)
WHERE qux = 1
```

## Steps to Reproduce
`sqlfluff fix`

## Version
Include the output of `sqlfluff --version` along with your Python version
sqlfluff, version 0.4.0a3
Python 3.7.4

## Configuration
```
Include your SQLFluff configuration here
```

### PR DESCRIPTION
**Pull request by alanmcruickshank**
This resolves #734 (or at least part of it).

This fix for the original query is now:

```sql
SELECT
    *
FROM foo
INNER JOIN bar
    USING (baz)
WHERE qux = 1
```

however, if the `bar` is on a new line it still fixes to:

```sql
SELECT
    *
FROM foo
INNER JOIN
        bar
    USING (baz)
WHERE qux = 1
```

I've also tried to add a load more logging in. @barrywhart - does that also make the structure easier to understand?

### CODE REVIEW

**Comment by alanmcruickshank** [comment]
We might want to revisit the latter problem once #882 is merged, because I think that might make things easier - and might also fix it.

**Comment by alanmcruickshank** [comment]
BOOM 🎉 

Now that #882 is merged it also fixes the indentation of `bar`:

```sql
SELECT
    *
FROM foo
INNER JOIN
    bar
    USING (baz)
WHERE qux = 1
```

**Review by barrywhart — COMMENTED** [review]
The code changes look good. A couple of existing tests broke, and maybe we want to add a test or two for the new behavior?

**Comment by barrywhart** [comment]
Some thoughts on other things that could make L003 easier to understand and modify:
* The concept of `clean_indent` seems pretty important to the design, but is hard to grasp from the existing comments. It might be helpful to add test cases to `_process_raw_stack()`, potentially using pytest "[parametrize](https://docs.pytest.org/en/stable/parametrize.html)", to explain by example
* Maybe the same as above but for detecting "hanging indent"
* It could be helpful to create smaller tests for the same behavior as the two large tests, but finer grained
  * `test/core/rules/std_fix_auto_test.py:143 test__std_fix_auto[ansi-002_indentation]`
  * `test/core/rules/std_fix_auto_test.py:143 test__std_fix_auto[bigquery-004_templating]`

**Comment by alanmcruickshank** [comment]
@barrywhart - can I get a re-review? I've added a load of test cases and commenting to improve coverage.

Fixed a few bugs too in the process. 🚀 

The `bigquery` templating example is still behaving a little strange, but I think that's more to do with the templating than L003.

File: test/core/rules/std_L003_test.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+from sqlfluff.core.config import FluffConfig
+from sqlfluff.core.rules.std.L003 import Rule_L003
+from sqlfluff.testing.rules import get_rule_from_set
+
+
+@pytest.mark.parametrize(
+    "test_elems,result",
+    [
+        # Unindented test examples
```
Comment (barrywhart) [inline test/core/rules/std_L003_test.py:13]: I 💕 these detailed test cases!

File: test/core/rules/test_cases/L003.yml [hunk 2]
```diff
@@ -1,24 +1,26 @@
...
+  # Comments should be aligned to the following line.
   pass_str: |
     SELECT
         -- Compute the thing
         (a + b) AS c
     FROM
         acceptable_buckets
 
-test_4:
+test__identation_of_comments_2:
```
Comment (barrywhart) [inline test/core/rules/test_cases/L003.yml:22]: Typos:
* Two underscores after "test"
* identation -> indentation

File: test/core/rules/test_cases/L003.yml [hunk 3]
```diff
@@ -167,3 +165,45 @@ bigquery_with_indent:
...
+                  bar)
+       FROM tbl
+
+  fix_str: |
+    SELECT coalesce(foo,
+                    bar)
+    FROM tbl
+
+test_possible_hanger_fix:
+  # It's not even close to a hnaging indent, but we should
```
Comment (barrywhart) [inline test/core/rules/test_cases/L003.yml:182]: Typo: hnaging -> hanging

File: test/core/rules/test_cases/L003.yml [hunk 4]
```diff
@@ -1,24 +1,26 @@
 rule: L003
 
-test_1:
```
Comment (barrywhart) [inline test/core/rules/test_cases/L003.yml:3]: Great tests! Love the new ones, the names, and the comments. My 🧠 is so happy now, ha ha.

File: test/fixtures/linter/autofix/bigquery/004_templating/after.sql [hunk 5]
```diff
@@ -24,6 +24,8 @@ raw_effect_sizes AS (
     SELECT
         COUNT(1) AS campaign_count_{{action}},
         {{corr_states}}
+    -- NOTE: The L003 fix routine behaves a little strangely here around the templated
```
Comment (barrywhart) [inline test/fixtures/linter/autofix/bigquery/004_templating/after.sql:27]: Can you clarify what is the strange behavior seen here? Is it the indentation of the `STDDEV_POP(` line?

File: test/fixtures/linter/autofix/bigquery/004_templating/before.sql [hunk 6]
```diff
@@ -24,6 +24,8 @@ WITH
   SELECT
     COUNT(1) AS campaign_count_{{action}},
     {{corr_states}}
+    -- NOTE: The L003 fix routine behaves a little strangely here around the templated
```
Comment (barrywhart) [inline test/fixtures/linter/autofix/bigquery/004_templating/before.sql:27]: Also here.

**Review by barrywhart — APPROVED** [review]
This is great! 🎉💯

I found a few typos, and there are a couple of places where comments could be clearer. Feel free to merge once you're ready; I don't see a need for any additional review.

**Pull request merged**